### PR TITLE
[ckpt] feat: Added an option to save only the top checkpoints in FSDP SFT

### PR DESCRIPTION
**Pull request by AJain9199**
### What does this PR do?

> Add **concise** overview of what this PR aims to achieve or accomplish. Reference related GitHub issues and PRs that help with the review.

Adds the `save_best` option to the FSDP SFT trainer config, which makes the trainer save only the `max_ckpt_to_keep` best checkpoints (based on val loss) or the single best checkpoint if `max_ckpt_to_keep` is not set.

If set, the `save_freq` must be a multiple of `test_freq`, so that there is an up-to-date `val_loss` which can be used to determine if the current model should be saved.

[This issue](https://github.com/volcengine/verl/issues/3792) sums up the rationale behind the feature.

In addition, it also logs the `grad_norm`. This makes it possible to monitor gradients and observe gradient explosions during training.

### Checklist Before Starting
- [x] Search for similar PRs. Paste at least one query link here: 
[This PR](https://github.com/volcengine/verl/pull/2173) suggested this feature for the PPO pipeline but wasn't accepted.
- [x] Format the PR title as `[{modules}] {type}: {description}` (This will be checked by the CI)
  - `{modules}` include `fsdp`, `megatron`, `sglang`, `vllm`, `rollout`, `trainer`, `ci`, `training_utils`, `recipe`, `hardware`, `deployment`, `ray`, `worker`, `single_controller`, `misc`, `perf`, `model`, `algo`, `env`, `tool`, `ckpt`, `doc`, `data`, `cfg`, `reward`
  - If this PR involves multiple modules, separate them with `,` like `[megatron, fsdp, doc]`
  - `{type}` is in `feat`, `fix`, `refactor`, `chore`, `test`
  - If this PR breaks any API (CLI arguments, config, function signature, etc.), add `[BREAKING]` to the beginning of the title.
  - Example: `[BREAKING][fsdp, megatron] feat: dynamic batching`

### Test

> For changes that can not be tested by CI (e.g., algorithm implementation, new model support), validate by experiment(s) and show results like training curve plots, evaluation results, etc.

```
torchrun -m verl.trainer.fsdp_sft_trainer ... trainer.save_best=True trainer.max_ckpt_to_keep=2
```

In an experiment run, with the following val_loss plot:
Screenshot 2025-12-26 204142

Only the checkpoints at 30 and 32 are saved.

### API and Usage Example

> Demonstrate how the API changes if any, and provide usage example(s) if possible.

An optional `current_val_loss` parameter is added to `FSDPCheckpointManager.save_checkpoint`. `save_best` functionality is used if the param is not `None`. Should not be breaking as it is optional
```python
# verl/utils/checkpoint/fsdp_checkpoint_manager.py
class FSDPCheckpointManager(BaseCheckpointManager):
# ...
    def save_checkpoint(
        self, local_path: str, hdfs_path: str = None, global_step: int = 0, max_ckpt_to_keep=None, current_val_loss=None
    )
```

### Design & Code Changes

> Demonstrate the high-level design if this PR is complex, and list the specific changes.

### Checklist Before Submitting

> [!IMPORTANT]
> Please check all the following items before requesting a review, otherwise the reviewer might deprioritize this PR for review.

- [x] Read the [Contribute Guide](https://github.com/volcengine/verl/blob/main/CONTRIBUTING.md).
- [x] Apply [pre-commit checks](https://github.com/volcengine/verl/blob/main/CONTRIBUTING.md#code-linting-and-formatting): `pre-commit install && pre-commit run --all-files --show-diff-on-failure --color=always`
- [ ] Add / Update [the documentation](https://github.com/volcengine/verl/tree/main/docs).
- [ ] Add unit or end-to-end test(s) to [the CI workflow](https://github.com/volcengine/verl/tree/main/.github/workflows) to cover all the code. If not feasible, explain why: ...
- [ ] Once your PR is ready for CI, send a message in [the `ci-request` channel](https://verl-project.slack.com/archives/C091TCESWB1) in [the `verl` Slack workspace](https://join.slack.com/t/verl-project/shared_invite/zt-3855yhg8g-CTkqXu~hKojPCmo7k_yXTQ). (If not accessible, please try [the Feishu group (飞书群)](https://applink.larkoffice.com/client/chat/chatter/add_by_link?link_token=772jd4f1-cd91-441e-a820-498c6614126a).)

### CODE REVIEW

File: verl/trainer/fsdp_sft_trainer.py [hunk 1]
```diff
@@ -786,15 +798,17 @@ def fit(self):
                         )
                         val_loss = self.validation_step(val_data)
                         val_losses.append(val_loss)
+
+                    val_loss = torch.mean(torch.stack(val_losses))
                     if rank == 0:
-                        val_loss = torch.mean(torch.stack(val_losses))
                         metric = {"val/loss": val_loss.detach().item()}
                         tracking.log(data=metric, step=global_step)
                         last_valid_metric = metric
```
[the remark this answers is not in the document]
Comment (AJain9199) [reply verl/trainer/fsdp_sft_trainer.py:809]: shouldn't be an issue. the preexisting code requires a val_dataloader

**Comment by vermouth1992** [comment]
This script is going to be deprecate. Please use sft_trainer
